controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                        return;
+                    }
+                };
+                debug!(log, "received outgoing request: {request:?}");
+
+                // Store the outstanding request, sanity-checking that we really
+                // didn't have a prior one.
+                let old = outstanding_request.replace(request);
+                assert!(
+                    old.is_none(),
+                    "dequeued a new request while one is already outstanding!",
+                );
+
+                let request = outstanding_request.as_mut().unwrap();
+                match send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await {
+                    Ok(n) => debug!(log, "sent message"; "message" => ?request, "n_bytes" => n),
+                    Err(e) => error!(log, "failed to send message"; "reason" => ?e),
+                }
+            }
+
+            // If we _do_ have an outstanding request, we need to resend it
+            // periodically until we get a response. Wait for up to the resend
+            // interval of inactivity, and then possibly retry.
+            _ = retry_timeout, if outstanding_request.is_some() => {
+                debug!(log, "timed out without response, retrying");
+                let request = outstanding_request.as_mut().unwrap();
+                send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await.unwrap();
+                info!(log, "resent message");
+            }
+
+            // Poll for outgoing responses we need to send.
+            maybe_response = outgoing_response_rx.recv() => {
+                let response = match maybe_response {
+                    Some(r) => r,
+                    None => {
+                        debug!(log, "outgoing response channel closed, exiting");
+                        return;
+                    }
+                };
+                // TODO-implement
+                debug!(log, "outgoing response: {response:?}");
+            }
+
+            // Poll for incoming packets.
+            res = socket.recv_from(&mut recv_buf) => {
+                let (n_bytes, peer) = match res {
+                    Err(e) => {
+                        error!(log, "I/O error receiving UDP packet: {e:?}");
+                        continue;
+                    }
+                    Ok((n_bytes, peer)) => (n_bytes, peer),
+                };
+
+                // Deserialize the message itself.
+                let (message, remainder): (Message, _) = match hubpack::deserialize(&recv_buf) {
+                    Err(e) => {
+                        error!(
+                            log,
+                            "failed to deserialize message";
+                            "reason" => ?e,
+                            "peer" => peer,
+                            "n_bytes" => n_bytes,
+                        );
+                        continue;
```

Comment:
Per [this discussion](https://github.com/oxidecomputer/hubris/pull/931#discussion_r1022991928), we should try deserializing just the `Header` if deserializing the whole message fails.

In the case of the SP sending a `HostRequest` with a mismatched version, deserializing the header will let us use the correct `message_id` in our `Error` reply.

Reply:
I'm actually not sure I understand. If we can't deserialize the message, the only thing we can check is the version. We don't know what kind of message it was supposed to be, so we can't use that to decide how to respond. We can still send a protocol error, but we can't do something like decide whether this was a request.

So maybe a hierarchy like this would work:

- If the message fails to deserialize, either drop the message or send a protocol error (or make a specific error for this).
- If the message serializes but is of the wrong type, send a protocol error.
- If the message is of the right type but wrong version, send a version mismatch error.

Reply:
That's a good point – I forgot that we can't decode the message _type_ from the header alone.

I still think we should try decoding the header, so that we can **log** a version mismatch, but think that we should drop packets with version mismatches (regardless of whether deserialization succeeds or not): if the version is mismatched, then even if deserialization "succeeds", we can't necessarily trust the data.

(also, if we replied, I could see two devices with different versions getting into a loop of spamming version mismatch errors at each other)

Reply:
> That's a good point – I forgot that we can't decode the message type from the header alone.

Another option would be to include the type as a separate enum in the header, but that feels a bit duplicative.

> I still think we should try decoding the header, so that we can log a version mismatch, but think that we should drop packets with version mismatches (regardless of whether deserialization succeeds or not): if the version is mismatched, then even if deserialization "succeeds", we can't necessarily trust the data.

Good point. That seems like a recipe for deserializing the _wrong_ thing, rather than just failing to deserialize.

To summarize, the plan is to:

- If deserializing the message fails, deserialize the header. Log the failure and then drop the packet.
- If deserializing the message succeeds, but there is a version mismatch, log and drop the packet. This is regardless of the _type_ of message, on the theory that we may not be able to accurately tell whether the message violates the protocol. Do _not_ send a version mismatch error.
- If deserializing the message succeeds _and_ the version is fine, check that it satisfies the protocol. If not, send a protocol error.

> (also, if we replied, I could see two devices with different versions getting into a loop of spamming version mismatch errors at each other)

As hilarious as that would be, we should probably avoid it :)